### PR TITLE
reset received vars on pserver

### LINKED ISSUE #13163: se-resnext多机同步下，有trainer在多个pass下后，loss值突增
paddle version：0.15
模型：se-resnext
paddlecloud job:
http://paddlecloud.baidu-int.com:8088/paddle/jobRunInfo?jobId=job-e6c5b7d5391e215f&flag=jobs&groupName=k8s_gpu_demo&groupId=c0a1f165-6279-5320-b9e7-e0218c7a87f5&currentPage=1&currentKey=1

### PR DESCRIPTION
**Pull request by Yancey0623**
fixed #13163

### CODE REVIEW

File: paddle/fluid/operators/listen_and_serv_op.cc [hunk 1]
```diff
@@ -166,15 +168,42 @@ void ListenAndServOp::RunSyncLoop(
...
 
     rpc_service_->SetCond(distributed::kRequestGet);
     rpc_service_->WaitBarrier(distributed::kRequestGet);
     rpc_service_->ResetBarrierCounter();
   }  // while(true)
 }
 
+void ListenAndServOp::ResetReceivedVars(
+    const std::vector<std::string> &recv_varnames, framework::Scope *recv_scope,
+    platform::DeviceContext *dev_ctx, bool only_sparse_vars) const {
```
Comment (typhoonzero) [inline paddle/fluid/operators/listen_and_serv_op.cc]: only_sparse_vars -> reset_all

File: paddle/fluid/operators/listen_and_serv_op.cc [hunk 2]
```diff
@@ -166,15 +168,42 @@ void ListenAndServOp::RunSyncLoop(
...
+  for (auto &varname : recv_varnames) {
+    auto var = recv_scope->FindVar(varname);
+    if (var == nullptr) {
+      VLOG(2) << "can not find var " << varname << " in received scope";
+      continue;
+    }
+    if (var->IsType<framework::SelectedRows>()) {
+      var->GetMutable<framework::SelectedRows>()->mutable_rows()->clear();
+    }
+    if (!only_sparse_vars) {
```
Comment (typhoonzero) [inline paddle/fluid/operators/listen_and_serv_op.cc]: Use `UNLIKLY(xxx)` to speedup for normal case

File: paddle/fluid/operators/listen_and_serv_op.cc (hunk 1 above)
Comment (Yancey0623) [reply paddle/fluid/operators/listen_and_serv_op.cc]: Done.

File: paddle/fluid/operators/listen_and_serv_op.cc (hunk 2 above)
Comment (Yancey0623) [reply paddle/fluid/operators/listen_and_serv_op.cc]: Done.

**Review by typhoonzero — DISMISSED** [review]

File: paddle/fluid/operators/listen_and_serv_op.cc [hunk 3]
```diff
@@ -165,16 +167,44 @@ void ListenAndServOp::RunSyncLoop(
...
     rpc_service_->SetCond(distributed::kRequestGet);
     rpc_service_->WaitBarrier(distributed::kRequestGet);
     rpc_service_->ResetBarrierCounter();
   }  // while(true)
 }
 
+void ListenAndServOp::ResetReceivedVars(
+    const std::vector<std::string> &recv_varnames, framework::Scope *recv_scope,
+    platform::DeviceContext *dev_ctx, bool reset_all) const {
+  for (auto &varname : recv_varnames) {
```
Comment (typhoonzero) [inline paddle/fluid/operators/listen_and_serv_op.cc]: Should optimize the performance here.

Comment (Yancey0623) [reply paddle/fluid/operators/listen_and_serv_op.cc]: Done.

File: paddle/fluid/operators/listen_and_serv_op.cc [hunk 4]
```diff
@@ -258,6 +313,7 @@ void ListenAndServOp::RunImpl(const framework::Scope &scope,
 
   bool sync_mode = Attr<bool>("sync_mode");
   auto fan_in = Attr<int>("Fanin");
+  auto inputs = Inputs("X");
```
Comment (jacquesqiao) [inline paddle/fluid/operators/listen_and_serv_op.cc:316]: Some variable may not in the inputs, need to check.

File: paddle/fluid/operators/listen_and_serv_op.cc [hunk 5]
```diff
@@ -165,16 +167,50 @@ void ListenAndServOp::RunSyncLoop(
...
+  if (UNLIKELY(reset_all)) {
+    for (auto &varname : dense_vars_) {
+      auto var = recv_scope->FindVar(varname);
+      if (var == nullptr) {
+        VLOG(2) << "can not find var " << varname << " in received scope";
+        continue;
+      }
+      if (var->IsType<framework::LoDTensor>()) {
+        math::set_constant(*dev_ctx, var->GetMutable<framework::LoDTensor>(),
+                           static_cast<float>(0));
```
Comment (jacquesqiao) [inline paddle/fluid/operators/listen_and_serv_op.cc:203]: Some tensor maybe is not `float`.

Comment (Yancey0623) [reply paddle/fluid/operators/listen_and_serv_op.cc:203]: The data type is according to the data type of the Tensor, the implement with CPU Place is as follows:

https://github.com/PaddlePaddle/Paddle/blob/a39eba77eb44f7b56159b5ae71bb37760f955a83/paddle/fluid/operators/math/math_function.cc#L68-L73

Comment (jacquesqiao) [reply paddle/fluid/operators/listen_and_serv_op.cc:203]: ok

File: paddle/fluid/operators/listen_and_serv_op.cc (hunk 4 above)
Comment (Yancey0623) [reply paddle/fluid/operators/listen_and_serv_op.cc:316]: Seems all the received gradients should be in the input arguments, don't double check the input vars in each iteration can archive higher performance.


https://github.com/PaddlePaddle/Paddle/blob/a39eba77eb44f7b56159b5ae71bb37760f955a83/python/paddle/fluid/transpiler/distribute_transpiler.py#L497-L500

Comment (Yancey0623) [reply paddle/fluid/operators/listen_and_serv_op.cc:316]: The gradients of dist lookup table would also been added into the received list:

https://github.com/PaddlePaddle/Paddle/blob/2fd1bf2ea6b6e0d27fb49461dd2b35c8e2a2b13b/python/paddle/fluid/transpiler/distribute_transpiler.py#L890-L891

https://github.com/PaddlePaddle/Paddle/blob/2fd1bf2ea6b6e0d27fb49461dd2b35c8e2a2b13b/python/paddle/fluid/transpiler/distribute_transpiler.py#L920-L925

**Review by typhoonzero — APPROVED** [review]

**Pull request merged**